File: go/internal/models/jsonfield.go
// Package models defines the storefront's domain types and the JSON-string-field convention
// Mudbase Collections require for anything array/object shaped.
package models

import "encoding/json"

// ParseJSONField decodes a JSON-encoded string field into T, returning fallback on an empty or
// malformed value. Mudbase Collection fields have a fixed type enum (string, number, boolean,
// date, email, url, enum, reference) with no native array/object type, so order line items,
// shipping addresses, and extra product image URLs are stored JSON-encoded in a string field and
// parsed at the edges - a documented constraint of the platform's Collections feature, not a
// workaround specific to this app (see web/README.md "Known limitations").
func ParseJSONField[T any](value string, fallback T) T {
	if value == "" {
		return fallback
	}
	var out T
	if err := json.Unmarshal([]byte(value), &out); err != nil {
		return fallback
	}
	return out
}

// StringifyJSONField encodes a value as a JSON string for storage in a Mudbase string field.
func StringifyJSONField[T any](value T) (string, error) {
	b, err := json.Marshal(value)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

File: python/app/json_field.py
"""JSON-string field helpers.

Mirrors web/src/lib/json-field.ts: Mudbase Collection fields have a fixed type
enum (string, number, boolean, date, email, url, enum, reference) with no
native array/object type. Anything shaped like a list or a nested record
(order line items, a shipping address, extra product images) is stored as a
JSON string in a `string` field and parsed at the edges. Real, documented
platform constraint — not a workaround specific to this app.
"""

import json
from typing import Any, TypeVar, cast

T = TypeVar("T")


def parse_json_field(value: str | None, fallback: T) -> T:
    """Parses a JSON-encoded string field, falling back on missing/invalid input.

    `T` is inferred from `fallback`; the parsed JSON is trusted to match it
    (the shape is controlled entirely by this app's own writes) rather than
    validated here — callers that build a Pydantic model from the result
    (e.g. `Order.items`) get that validation for free at that point instead.
    """
    if not value:
        return fallback
    try:
        return cast(T, json.loads(value))
    except (ValueError, TypeError):
        return fallback


def stringify_json_field(value: Any) -> str:
    """Encodes a value for storage in a JSON-string collection field."""
    return json.dumps(value)
